spark/src/main/scala/org/apache/spark/sql/delta/schema/ImplicitMetadataOperation.scala

```diff
@@ -236,42 +337,23 @@ object ImplicitMetadataOperation {
       protocol: Protocol,
       metadata: actions.Metadata,
       dataSchema: StructType): Unit =
-  SchemaMergingUtils.transformColumns(metadata.schema, dataSchema) {
-    case (fieldPath, currentField, Some(updateField), _)
-      // This condition is actually too strict, structs may be identified as changing because one
-      // of their field is changing even though that field isn't referenced by any constraint or
-      // generated column. This is intentional to keep the check simple and robust, esp. since it
-      // aligns with the historical behavior of this check.
-      if !SchemaMergingUtils.equalsIgnoreCaseAndCompatibleNullability(
-        currentField.dataType,
-        updateField.dataType
-      ) =>
-        val columnPath = fieldPath :+ currentField.name
-        // check if the field to change is referenced by check constraints
-        val dependentConstraints =
-          Constraints.findDependentConstraints(sparkSession, columnPath, metadata)
-        if (dependentConstraints.nonEmpty) {
-          throw DeltaErrors.constraintDataTypeMismatch(
-            columnPath,
-            currentField.dataType,
-            updateField.dataType,
-            dependentConstraints
-          )
-        }
-        // check if the field to change is referenced by any generated columns
-        val dependentGenCols = SchemaUtils.findDependentGeneratedColumns(
-          sparkSession, columnPath, protocol, metadata.schema)
-        if (dependentGenCols.nonEmpty) {
-          throw DeltaErrors.generatedColumnsDataTypeMismatch(
-            columnPath,
-            currentField.dataType,
-            updateField.dataType,
-            dependentGenCols
-          )
-        }
-      // We don't transform the schema but just perform checks, the returned field won't be used
-      // anyway.
-      updateField
-    case (_, field, _, _) => field
-  }
+    SchemaMergingUtils.transformColumns(metadata.schema, dataSchema) {
+      case (fieldPath, currentField, Some(updateField), _)
+        if !SchemaMergingUtils.equalsIgnoreCaseAndCompatibleNullability(
+          currentField.dataType,
+          updateField.dataType
+        ) =>
+        checkReferencedByCheckConstraintsOrGeneratedColumns(
```

Comment:
`SchemaMergingUtils.transformColumns` will apply the transform to all StructField in the schema, so you will end up calling `checkReferencedByCheckConstraintsOrGeneratedColumns` multiple times for nested struct fields since that method also recurses into the schema.

For example:
`col struct<a: struct<x: int, y: int>>`
`transformColumns` will call `checkReferencedByCheckConstraintsOrGeneratedColumns` on `col` first, which will recurse into the struct and call itself on `a`.
Then `transformColumns` will call `checkReferencedByCheckConstraintsOrGeneratedColumns` again on `a` directly.

Not really an issue, you can either skip recursing into structs in `checkReferencedByCheckConstraintsOrGeneratedColumns` or call out in a comment there that it's redundant

Reply:
thanks for pointing this out!

I've skipped the recursing into structs for `checkReferencedByCheckConstraintsOrGeneratedColumns`, will finalize this after the discussion regarding `MapType` and `ArrayType` [here](https://github.com/delta-io/delta/pull/3601#discussion_r1732091240).